src/rust/engine/src/intrinsics.rs

```diff
@@ -786,6 +786,7 @@ fn parse_python_deps(context: Context, args: Vec<Value>) -> BoxFuture<'static, N
         let cache_key = CacheKey {
           key_type: CacheKeyType::DepInferenceRequest.into(),
           digest: Some(digest.into()),
+          metadata: metadata,
```

Comment:
```diff
--- a/src/rust/engine/protos/protos/pants/cache.proto
+++ b/src/rust/engine/protos/protos/pants/cache.proto
@@ -15,7 +15,11 @@ message CacheKey {
   CacheKeyType key_type = 1;
 
   build.bazel.remote.execution.v2.Digest digest = 2;
-  optional string metadata = 3;  // Any other metadata to accompany the tag
+  optional build.bazel.remote.execution.v2.Digest metadata = 3;  // Any other metadata to accompany the tag
+}
+
+message DependencyInferenceMetadata {
+  build.bazel.remote.execution.v2.Digest metadata = 1;
 }

```
Is this what you're suggesting?

Reply:
```diff
@@ -15,7 +15,11 @@ message CacheKey {
   CacheKeyType key_type = 1;
 
   build.bazel.remote.execution.v2.Digest digest = 2;
-  optional string metadata = 3;  // Any other metadata to accompany the tag
+}
+
+message DependencyInferenceRequest {
+  build.bazel.remote.execution.v2.Digest digest = 1;
+  optional build.bazel.remote.execution.v2.Digest metadata = 2;  // Any other metadata to accompany the tag
 }

```
Or this?

Reply:
Right, except you can actually give the `digest` in `DependencyInferenceRequest` a name: it's the input file digest. You can probably also use a more specific type for metadata, and/or let it actually be typed as multiple fields.

Reply:
Then it would make sense to me to just expose the metadata types for any inference impl via pyo3 instead of json, but that conflicts with comment/want this a bit:
https://github.com/pantsbuild/pants/pull/18985#issuecomment-1546643872

Reply:
Since metadata is per impl (I know JS and TS wont share, anyway), is a `one of`  "union" appropriate? I have no exp with  protocol buffers.

Reply:
In that case, making `metadata` a [map](https://protobuf.dev/programming-guides/proto3/#maps) seems fine, although putting any implementation agnostic fields outside of it would be good.

Reply:
A map cannot contain maps, the structure Im passing right now is
```json
{
"root": "some/dir", 
"patterns": {"a-pat/*: ["replace-me/*", "and-me-*"],  ...<more-patterns>}}
```
Wouldnt a message per type fit better?

Reply:
I realize the whole patterns structure can be a message, but unsure if that is what you had in mind